src/createSettingsFileTransformer.js

```diff
@@ -0,0 +1,162 @@
+var objectAssign = require('@adobe/reactor-object-assign');
+
+function isArrayReference(str) {
+  return isString(str) && str.indexOf('[') !== -1 && str.indexOf(']') !== -1;
+}
+function sanitizeArrayKey(pathStrSegment) {
+  return pathStrSegment.substr(
+    0,
+    // the name goes up to the start of the array bracket: 'someArrayName[]'
+    pathStrSegment.indexOf('[')
+  );
+}
+
+function isObject(value) {
+  if (value == null) {
+    return false;
+  }
+
+  return Boolean(
+    typeof value === 'object' &&
+      Object.prototype.toString.call(value) === '[object Object]' &&
+      !Array.isArray(value)
+  );
+}
+
+function isString(value) {
+  var type = typeof value;
+  return Boolean(
+    type === 'string' ||
+      (type === 'object' &&
+        value != null &&
+        !Array.isArray(value) &&
+        Object.prototype.toString.call(value) === '[object String]')
+  );
+}
+
+function traverseIntoSettings(
+  filePathString,
+  settings,
+  decorateWithDynamicHost
+) {
+  // nothing to do
+  if (
+    !isString(filePathString) ||
+    !filePathString.length ||
+    !isObject(settings)
+  ) {
+    return;
+  }
+
+  // get the next key to observe
+  var pathSegments = filePathString.split('.');
```

Comment:
I think splitting the filePathString in an array just to join it back on line 76 is not optimal. I think we should split the path once in the function that calls traverseIntoSettings, and send as a first parameter an array with the segments. From my tests the difference between these two approaches can get very large. I did a test here and you can see the perf results: https://codesandbox.io/s/perf-wq9hk.